src/transport/CASESession.cpp

```diff
@@ -121,81 +270,287 @@ CHIP_ERROR CASESession::DeriveSecureSession(SecureSession & session)
+    err = SendSigmaR2();
+    SuccessOrExit(err);
 
-    System::PacketBufferHandle msg_R2;
+exit:
+    return err;
+}
 
-    ChipLogDetail(Inet, "Received SigmaR1 msg");
+CHIP_ERROR CASESession::HandleSigmaR1(const PacketHeader & header, const System::PacketBufferHandle & msg)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+
+    const uint8_t * buf   = msg->Start();
+    uint16_t buflen       = msg->DataLength();
+    uint16_t fixed_buflen = kSigmaParamRandomNumberSize + sizeof(uint16_t) + kTrustedRootIdSize + kP256_PublicKey_Length;
+
+    CertificateKeyId trustedRoot[kMaxTrustedRootIds];
 
     VerifyOrExit(buf != nullptr, err = CHIP_ERROR_MESSAGE_INCOMPLETE);
+    VerifyOrExit(buflen >= fixed_buflen, err = CHIP_ERROR_INVALID_MESSAGE_LENGTH);
+
+    ChipLogDetail(Inet, "Received SigmaR1 msg");
+
+    err = mCommissioningHash.AddData(msg->Start(), msg->DataLength());
+    SuccessOrExit(err);
+
+    {
+        // Let's skip the random number portion of the message
+        const uint8_t * msgptr = &buf[kSigmaParamRandomNumberSize];
+        // Omitting SessionID here - Mirroring PASESession, which omits SessionID as well
+        uint32_t n_trusted_roots = chip::Encoding::LittleEndian::Read16(msgptr);
+        // Step 1/2
+        for (uint32_t i = 0; i < n_trusted_roots; ++i)
+        {
+            trustedRoot[i].mId  = msgptr;
+            trustedRoot[i].mLen = kTrustedRootIdSize;
+            msgptr += kTrustedRootIdSize;
+
+            if (mOpCredSet->IsTrustedRootIn(trustedRoot[i]))
+            {
+                mTrustedRootId.mId = reinterpret_cast<const uint8_t *>(chip::Platform::MemoryAlloc(kTrustedRootIdSize));
+                VerifyOrExit(mTrustedRootId.mId != nullptr, err = CHIP_ERROR_NO_MEMORY);
+
+                memcpy(const_cast<uint8_t *>(mTrustedRootId.mId), trustedRoot[i].mId, trustedRoot[i].mLen);
+                mTrustedRootId.mLen = trustedRoot[i].mLen;
+
+                break;
+            }
+        }
+        VerifyOrExit(mTrustedRootId.mId != nullptr, err = CHIP_ERROR_CERT_NOT_TRUSTED);
+        // write public key from message
+        Encoding::LittleEndian::BufferWriter bbuf(&*mRemotePubKey, mRemotePubKey.Length());
+        bbuf.Put(msgptr, kP256_PublicKey_Length);
+        VerifyOrExit(bbuf.Fit(), err = CHIP_ERROR_NO_MEMORY);
+    }
 
-    // TODO: Verify received SigmaR1 message.
-    // https://github.com/project-chip/connectedhomeip/issues/4469
+    if (header.GetSourceNodeId().HasValue() && mConnectionState.GetPeerNodeId() == kUndefinedNodeId)
+    {
+        mConnectionState.SetPeerNodeId(header.GetSourceNodeId().Value());
+    }
 
     mConnectionState.SetPeerKeyID(header.GetEncryptionKeyID());
 
+exit:
+
+    if (err == CHIP_ERROR_CERT_NOT_TRUSTED)
+    {
+        SendErrorMsg(SigmaErrorType::kNoSharedTrustRoots);
+    }
+    else if (err != CHIP_NO_ERROR)
+    {
+        SendErrorMsg(SigmaErrorType::kUnexpected);
+    }
+    return err;
+}
+
+CHIP_ERROR CASESession::SendSigmaR2()
+{
```

Comment:
This is ~150 lines, with 15 local variables on top and an extra 5 variables within scopes. 

Could we split things up and maybe add more focused unit tests for various steps? I find hard to review and maintain such large pieces of code.